### ISSUE TITLE
tf.data.Dataset from_generator w/ keras.utils.Sequence doesn't trigger on_epoch_end

### ISSUE THREAD

**Issue by TheEnigmist** [tensorflow/tensorflow#48661]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow version (use command below): 2.4.1
- Python version: 3.8
- CUDA/cuDNN version: 11.0 / 8.0.4
- GPU model and memory: RTX 3070

**Describe the current behavior**
I created a class from `tf.keras.util.Sequence` with `on_epoch_end()` function with a print inside it. Then I created a `tf.data.Dataset from_generator` using that class. Finally used that dataset as input for my `model.fit()`.

**Describe the expected behavior**
I expect the print inside `on_epoch_end()` appear on console, but it doesn't so I suppose it never trigger that function. 

**Standalone code to reproduce the issue**
Generator class:
```
class HDF5Generator(tf.keras.utils.Sequence):
    def __init__(self, hdf5_file):
        print("GENERATED")
        self.hdf5 = h5py.File(hdf5_file, 'r')
        self.indices = list(range(0, len(self.hdf5["samples"])))
        random.Random().shuffle(self.indices)

    def __len__(self):
        return len(self.hdf5["samples"])

    def __getitem__(self, idx):
        return self.hdf5["samples"][self.indices[idx]], self.hdf5["labels"][self.indices[idx]]

    def on_epoch_end(self):
        print("SHUFFLE")
        random.Random().shuffle(self.indices)
```

Used class in fit
```
d = tf.data.Dataset.from_generator(HDF5Generator, args=[dataset], output_signature=(...))
d = d.batch(batch_size).prefetch(tf.data.AUTOTUNE).cache()
model.fit(d, epochs=epochs)
```
This is the output:
```
Epoch 1/40
GENERATED
1969/1969 [==============================] - 63s 29ms/step - loss: 0.4823 - accuracy: 0.8020 - f1_score: 0.7989
Epoch 2/40
1969/1969 [==============================] - 22s 11ms/step - loss: 0.1111 - accuracy: 0.9562 - f1_score: 0.9567
Epoch 3/40
1969/1969 [==============================] - 23s 12ms/step - loss: 0.0947 - accuracy: 0.9610 - f1_score: 0.9614
Epoch 4/40
```
As you can see no SHUFFLE string appear in output

**Comment by tilakrayal** [comment]
@TheEnigmist ,

I ran the code shared and face a different error, please find the [gist](https://colab.research.google.com/gist/tilakrayal/6e57541cc1214c25ace3f7de6eeb7cb0/48661.ipynb) here and share all dependencies to replicate the issue or share a colab gist with the reported error.

Thanks!

**Comment by TheEnigmist** [comment]
> @TheEnigmist ,
> 
> I ran the code shared and face a different error, please find the [gist](https://colab.research.google.com/gist/tilakrayal/6e57541cc1214c25ace3f7de6eeb7cb0/48661.ipynb) here and share all dependencies to replicate the issue or share a colab gist with the reported error.
> 
> Thanks!

Sorry, I copied only part of my code. I edited your gist so now you can run it. 
As you can see there is no output at the end of the epoch. I can put a raise Exception too and will never trigger!

**Comment by tilakrayal** [comment]
@TheEnigmist ,

I do not have access to the link you have provided. Could you please provide the required permissions to view the files.

Thanks!

**Comment by TheEnigmist** [comment]
> @TheEnigmist ,
> 
> I do not have access to the link you have provided. Could you please provide the required permissions to view the files.
> 
> Thanks!

Added view permission for the code. For those who don't want to access colab, the sample code is:

```
import tensorflow as tf
import numpy as np
import random
from tensorflow.keras import models
from tensorflow.python.keras.layers import Conv1D, Dense, MaxPooling1D, Flatten
from tensorflow import keras

class HDF5Generator(tf.keras.utils.Sequence):
    def __init__(self):
        print("GENERATED")
        self.samples = np.random.rand(10000,5,20)
        self.labels = np.random.randint(2, size=(10000, 1))
        self.indices = list(range(0, 10000))
        random.Random().shuffle(self.indices)

    def __len__(self):
        return 10000

    def __getitem__(self, idx):
        return self.samples[self.indices[idx]], self.labels[self.indices[idx]]

    def on_epoch_end(self):
        print("SHUFFLE")
        raise Exception("AAA")
        random.Random().shuffle(self.indices)

d = tf.data.Dataset.from_generator(HDF5Generator, output_signature=(tf.TensorSpec(shape=(5,20)), tf.TensorSpec(shape=(1,))))
d = d.batch(32).prefetch(tf.data.AUTOTUNE).cache()

model = models.Sequential()
model.add(Conv1D(32, kernel_size=1, strides=1, activation="relu", padding="same", input_shape=(5,20)))
model.add(MaxPooling1D(pool_size=3, padding="same"))
model.add(Conv1D(64, kernel_size=1, strides=1, activation="relu", padding="same", input_shape=(5,20)))
model.add(MaxPooling1D(pool_size=3, padding="same"))
model.add(Flatten())
model.add(Dense(1024, activation="relu"))
model.add(Dense(1, activation="softmax"))
model.compile(
        optimizer=keras.optimizers.Adam(0.003),
        loss="categorical_crossentropy",
        metrics=["accuracy"],
    )
model.fit(d, epochs=5)
```

The output is:
```
Epoch 1/5
GENERATED
313/313 [==============================] - 3s 7ms/step - loss: 0.0000e+00 - accuracy: 0.4886
Epoch 2/5
313/313 [==============================] - 1s 3ms/step - loss: 0.0000e+00 - accuracy: 0.4886
Epoch 3/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0000e+00 - accuracy: 0.4886
Epoch 4/5
313/313 [==============================] - 1s 3ms/step - loss: 0.0000e+00 - accuracy: 0.4886
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.0000e+00 - accuracy: 0.4886
```

**Comment by tilakrayal** [comment]
@jvishnuvardhan ,

I was able to reproduce the issue in TF v2.4,2.5.0rc1 and nightly.Please find the [gist](https://colab.research.google.com/gist/tilakrayal/b2f26e0229ab391028177599c7a9965c/48661.ipynb) here.

**Comment by jvishnuvardhan** [comment]
@TheEnigmist Based on a good [tutorial](https://stanford.edu/~shervine/blog/keras-how-to-generate-data-on-the-fly) and [SO issue](https://stackoverflow.com/questions/62346965/input-pipeline-w-keras-utils-sequence-object-or-tf-data-dataset), i updated your code with one line to init `on_epoch_end` in the __init__ method.

With that modification, now the code throws Exception. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/b2e37b7cfb380af4cbfd73a6b315c4e3/48661.ipynb). Hope it helps. Thanks!

**Comment by TheEnigmist** [comment]
Well that trigger `on_epoch_end()` only once at init time, but will not trigger it again.

If you remove the raise you will see that "SHUFFLE" print occurs only at before first epoch but never at the end of each epoch, like should be.

**Comment by chunduriv** [comment]
@TheEnigmist, According to [this](https://github.com/tensorflow/tensorflow/issues/48661#issuecomment-838069103), your issue was resolved.Please feel free to close the issue if it is resolved? Thanks!

**Comment by TheEnigmist** [comment]
@chunduriv Yes, but this is a workaround, still it doesn't work like explained in the docs. 
The expected behaviour is that `on_epoch_end()` of the class `Sequence` is triggered without forcing it from a callback.
If you think this is like it works then I close this issue.

**Comment by tilakrayal** [comment]
Hi,
I tried to execute the code with the alternative approach and observed that it is working as expected. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/99b1d0f3cd72bd704d2aa3519f75e2c7/untitled2122.ipynb).

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
